compiler-plugin/src/main/kotlin/arrow/meta/quotes/MetaExtensions.kt

```diff
@@ -0,0 +1,315 @@
+
+/**
+ * ```
+ * """ { $statements } """.block
+ * ```
+ */
+fun Meta.blockExpression(
+  match: KtBlockExpression.() -> Boolean,
+  map: BlockExpression.(KtBlockExpression) -> Transform<KtBlockExpression>
+): ExtensionPhase =
+  quote(match, map) { BlockExpression(it) }
+
+/**
+ * ```
+ * """ catch $`(parameter)` $`{ body }` """.catchClause
+ * ```
+ */
+fun Meta.catchClause(
+  match: KtCatchClause.() -> Boolean,
+  map: CatchClause.(KtCatchClause) -> Transform<KtCatchClause>
+): ExtensionPhase =
+  quote(match, map) { CatchClause(it) }
+
+/**
+ * ```
+ *  """|$`@annotations` $kind $name $`(typeParameters)` $`(params)` : $supertypes"} { $body } """.`class`
+ *  ```
+ */
+fun Meta.`class`(
+  match: KtClass.() -> Boolean,
+  map: ClassDeclaration.(KtClass) -> Transform<KtClass>
+): ExtensionPhase =
+  quote(match, map) { ClassDeclaration(it) }
+
+/**
+ * ```
+ * """ $`(vars)` = $initializer """.destructuringDeclaration
+ * ```
+ */
+fun Meta.destructuringDeclaration(
+  match: KtDestructuringDeclaration.() -> Boolean,
+  map: DestructuringDeclaration.(KtDestructuringDeclaration) -> Transform<KtDestructuringDeclaration>
+): ExtensionPhase =
+  quote(match, map) { DestructuringDeclaration(it) }
+
+/**
+ * ```
+ * """ $packageDirective $importList $classes """.file
+ * ```
+ */
+fun Meta.file(
+  match: KtFile.() -> Boolean,
+  map: File.(KtFile) -> Transform<KtFile>
+): ExtensionPhase =
+  quote(match, map) { File(it) }
+
+/**
+ * ```
+ * """ $finally """.finally
+ * ```
+ */
+fun Meta.finallySection(
+  match: KtFinallySection.() -> Boolean,
+  map: FinallySection.(KtFinallySection) -> Transform<KtFinallySection>
+): ExtensionPhase =
+  quote(match, map) { FinallySection(it) }
+
+/**
+ * ```
+ * """ val $name = { $blockExpression }""".functionLiteral
+ * ```
+ */
+fun Meta.functionLiteral(
+  match: KtFunctionLiteral.() -> Boolean,
+  map: FunctionLiteralScope.(KtFunctionLiteral) -> Transform<KtFunctionLiteral>
+) : ExtensionPhase =
+  quote(match, map) { FunctionLiteralScope(it) }
+
+/**
+ * ```
+ * """ if $`(condition)` $then $`else` """.`if`
+ * ```
+ */
+fun Meta.ifExpression(
+  match: KtIfExpression.() -> Boolean,
+  map: IfExpression.(KtIfExpression) -> Transform<KtIfExpression>
+): ExtensionPhase =
+  quote(match, map) { IfExpression(it) }
+
+/**
+ * ```
+ * """ import $importPath.$importedName """.importDirective
+ * ```
+ */
+fun Meta.importDirective(
+  match: KtImportDirective.() -> Boolean,
+  map: ImportDirective.(KtImportDirective) -> Transform<KtImportDirective>
+) : ExtensionPhase =
+  quote(match, map) { ImportDirective(it) }
+
+/**
+ * ```
+ * """ { $`(params)` -> $bodyExpression } """.lambdaExpression
+ * ```
+ */
+fun Meta.lambdaExpression(
+  match: KtLambdaExpression.() -> Boolean,
+  map: LambdaExpression.(KtLambdaExpression) -> Transform<KtLambdaExpression>
+) : ExtensionPhase =
+  quote(match, map) { LambdaExpression(it) }
+
+/**
+ * ```
+ * """ $`@annotations` $modifier value """.`modifierList`
```

Comment:
Is $modifier a property of the scope or can any soft keyword be a modifier?
If this points to a ScopedList then it probably should be renamed to `modifiers`

Reply:
For whatever reason this is a singular KtModifer - perhaps it's for the whole list